Add TestClient

myusko opened the pull request:
PR is related to #117 

@RobertoPrevato Hey, I've added a very basic draft of the `TestClient`, when you have a chance, please take a look into the implementation, and please share your thoughts, and then I'll rewrite/polish some stuff.

Small example, how it works right now.

```python
from blacksheep.server.application import Application
from blacksheep.testing.client import TestClient
from blacksheep.server.responses import json

app = Application()


@app.route("/foo")
async def foo(request):
    print(request.query)
    return json({"foo": "bar"})


@app.route("/bar")
async def bar(request):
    print(request.headers)
    return {"bar": "foo"}



client = TestClient(app)
response_one = await client.get("/foo", headers={"content-type": "application/json"})
response_two = await client.get("/bar", query={"name": "foo and bar"})
```

RobertoPrevato commented on blacksheep/testing/client.py:
```diff
@@ -0,0 +1,64 @@
+
+class TestClient:
+    def __init__(self, app):
+        self.app = app
```
I would recommend to instantiate a single `TestSimulator` per TestClient, and maybe to abstract the exact implementation of the object that is responsible of sending web requests. Because then we could support this scenario:

* by default the TestClient would use the `TestSimulator` implementation
* by configuration, TestClient could use a different kind of object to send requests, for example to generate real HTTP requests against a running server, so that the user by configuration can test mocked web requests (enabling super fast tests), or real web requests (useful to test end-to-end integration with real implementations of ASGI servers and http clients)

Something like - high level example -:

````python
from abc import ABC, abstractmethod


class TestHTTPTransport:
    @abstractmethod()
    async def send_request(...) -> void:
        """Sends a web request to the application."""


class TestClient:
    def __init__(self, app, transport: Optional[TestHTTPTransport] = None):
        if transport is None:
             transport = _TestSimulator(app)
        self._transport = transport
    
    async def send_request(...):
        return self._transport.send_request(...)

    async def get(...):
         return await self._send_request(...)
```

myusko commented on blacksheep/testing/simulator.py:
```diff
@@ -0,0 +1,79 @@
+
+        request.scope = scope
+
+        if content is not None:
+            encoded_content = json.dumps(content).encode()
+            request.content = ASGIContent(MockReceive([encoded_content]))
```
@RobertoPrevato If we want to support JSON/multipart body in the request, we should probably create a class that will set proper content with the proper class.

e.g
```
from blacksheep.testing import Content

response = await client.post("/users", content=Content(mime_type="application/json", data={"username": "foo"}))
```

That's was my first thought, but not sure if this the best way.

RobertoPrevato replied:
There is a `JSONContent` class, that can be used for responses and requests in the client. The `ASGIContent` class was added when I modified the framework to be ASGI compatible, it should be needed only for incoming content from the ASGI web server.

Can you please try with the existing `JSONContent` class? It should work.

myusko replied:
Yeah, will do.

RobertoPrevato replied:
It just occurred to me that maybe it would be best if the `send_request` method accepted as input an `Optional[Content]`, and not assume that the user wants to handle JSON. For example it could be also Form content, Multipart, raw bytes.

Otherwise, the method signature might support different kinds of keywords (this is the approach of `requests` library), e.g. 

```python
def example(..., json=None, form=None, files=None, multipart=None, raw=None): ...
```

See: https://docs.python-requests.org/en/master/user/quickstart/#make-a-request

![image](https://user-images.githubusercontent.com/2576032/119880451-8f629180-bf2c-11eb-8294-1abf3aa29fea.png)

RobertoPrevato replied:
What do you think?

myusko replied:
Yeah, I already pushed changes, and now for content argument we're expecting a Content's child class.

e.g

`JSONContent, FormContent`, etc.

RobertoPrevato reviewed (commented):
When I coded the [client part](https://github.com/Neoteroi/BlackSheep/blob/main/blacksheep/client/session.py), I opted for keeping the functions' signatures clean, and expect simply an instance of `Content`, that can be anything then.

So for example:

```python
client.post("https://example.dev/api/cat", content=JSONContent({"name": "Foo"}))
```

I don't like the approach in `requests` too much, because then you would need to handle also `xml=` for XML content to be automatically serialized? And then with which encoding (UTF-8, ascii, etc.)? It looks cleaner to me to just keep the content abstracted from a specific encoding and mime type.

RobertoPrevato commented:
@myusko maybe it's best to discuss the details in another place, because comments here are hard to read 😄

myusko commented:
> When I coded the [client part](https://github.com/Neoteroi/BlackSheep/blob/main/blacksheep/client/session.py), I opted for keeping the functions' signatures clean, and expect simply an instance of `Content`, that can be anything then.
> 
> So for example:
> 
> ```python
> client.post("https://example.dev/api/cat", content=JSONContent({"name": "Foo"}))
> ```
> 
> I don't like the approach in `requests` too much, because then you would need to handle also `xml=` for XML content to be automatically serialized? And then with which encoding (UTF-8, ascii, etc.)? It looks cleaner to me to just keep the content abstracted from a specific encoding and mime type.

Yeah, I already pushed changes, and now for the content argument, we're expecting a Content's child classes.

e.g

JSONContent, FormContent, etc.

myusko commented:
@RobertoPrevato PR is ready for the first round of review.

- Added TestClient
- Added tests
- Removed `messages`, `scopes` modules from the tests.

RobertoPrevato commented:
@myusko, it looks **Wonderful**! 🎉  I only have a few recommendations - writing now

RobertoPrevato commented on blacksheep/testing/simulator.py:
```diff
@@ -0,0 +1,100 @@
+        return response
+
+    def _prepare_application(self):
+        if self.app._service_provider is None:
```
Otherwise, if you wish to keep the async pattern in the `TestClient`, we might consider making it an async context manager. The application in such case would be started in `__aenter__` and disposed at `__aexit__`.

myusko replied:
Thank you for the good point, it really helps!

Probably I will check both options, and then we can decide which one we want to use, right now I prefer `loop.run_until_complete`, but let's see.

myusko commented on blacksheep/testing/simulator.py:
```diff
@@ -0,0 +1,100 @@
+    method: str,
+    path: str,
+    headers: Optional[Dict[str, str]] = None,
+    query: Optional[Dict[str, str]] = b"",
```
Yeah, will fix it, thank you 🎉

myusko commented:
Seems like, we should add more tests, to satisfy coverage, and `TestSimulator` is a good candidate for it.

RobertoPrevato commented on tests/test_testing.py:
```diff
@@ -0,0 +1,157 @@
+
+@pytest.fixture
+def test_app():
+    return Application(show_error_details=True)
```
Since pytest supports asynchronous fixtures, maybe we could remove the calls to `_start_application` and use instead:

```python
@pytest.fixture
async def test_app():
    app = Application(show_error_details=True)
    await app.start()
    yield app
    await app.stop()  # necessary only if the app needs to close gracefully things when it stops, not in this specific case
```

myusko replied:
That's a good point, I've tried the same example as you wrote, but tests failing with the next error
```
TypeError: Cannot convert dict to blacksheep.messages.Response
```

An interesting thing that in the tests we receiving a started application, for some reason, if we don't start the application within a test, it fails, thoughts?

RobertoPrevato replied:
Absolutely! I made a mistake - we cannot start the application in a fixture because you need to register request handlers as part of the test arrangements _before_ starting the application! Your code is good as it is now and it cannot work otherwise.

RobertoPrevato replied:
note: I marked it again as "unresolved" only because the GitHub UI hides too resolved conversations too much and I think they are almost impossible to see. 😄

RobertoPrevato commented on blacksheep/testing/client.py:
```diff
@@ -0,0 +1,96 @@
+    ):
+        self._test_simulator = test_simulator or TestSimulator(app)
+
+    async def get(
```
Looks great! Can you please add also methods for HTTP Methods: `OPTIONS`, `HEAD`, `TRACE`? I think leaving out `CONNECT` is fine, but especially `OPTIONS` and `HEAD` are relevant to test CORS and caching.

myusko replied:
Will do

RobertoPrevato commented on blacksheep/testing/simulator.py:
```diff
@@ -0,0 +1,97 @@
+            - delete
+        Then you can define an own TestClient, with the custom logic.
+        """
+        pass
```
This `pass` statement can be removed, since the abstract method declaration has a docstring. This is also useful for code coverage report, since it gives a false negative for `pass`:

![image](https://user-images.githubusercontent.com/2576032/120888226-74acae80-c5f7-11eb-8b0c-20df78425f45.png)

RobertoPrevato commented on blacksheep/testing/helpers.py:
```diff
@@ -9,6 +9,8 @@ def get_example_scope(
+    """Returns mocked ASGI scope"""
+
     if "?" in path:
         raise ValueError("The path in ASGI messages does not contain query string")
```
Since you are moving this piece of code from `tests` under the main namespace, it is now analyzed for code coverage. The only missing line for this file is the `"?" in path` check (bringing down the coverage for this file to 89%). So we should either add a test to assert that `get_example_ scope` throws if the path contains a query string, or eventually replacing the line with an assertion:

```python
assert "?" not in path, "The path in ASGI messages does not contain query string"
```

myusko replied:
Nice catch! will do

RobertoPrevato reviewed (commented):
Looks great, @myusko! I only added a few minor comments, then let's merge this into the main branch. I looked into the code coverage, and I only found the two lines I commented in the review to correct the code coverage - they are like false positives.

The web framework is 100% tested. The only thing that currently is not fully covered is the Client, because apparently after upgrading to the most recent version of `httptools`, it never returns true for `should_keep_alive()`

![image](https://user-images.githubusercontent.com/2576032/120888529-04069180-c5f9-11eb-912c-202844b5acfd.png)

This is something we should look into, but let's postpone it for now.
